File: internal/service/ocr_svc.go
package service

import (
	"context"
	"fmt"
	"os"

	"github.com/zajca/zfaktury/internal/domain"
	"github.com/zajca/zfaktury/internal/service/ocr"
)

// ocrSupportedContentTypes defines which document types can be processed by OCR.
var ocrSupportedContentTypes = map[string]bool{
	"image/jpeg":      true,
	"image/png":       true,
	"application/pdf": true,
}

// OCRService handles OCR processing of expense documents.
type OCRService struct {
	provider  ocr.Provider
	documents *DocumentService
}

// NewOCRService creates a new OCRService.
func NewOCRService(provider ocr.Provider, documents *DocumentService) *OCRService {
	return &OCRService{
		provider:  provider,
		documents: documents,
	}
}

// ProcessDocument retrieves a document by ID, reads its file data,
// and sends it to the OCR provider for extraction.
func (s *OCRService) ProcessDocument(ctx context.Context, documentID int64) (*domain.OCRResult, error) {
	if documentID == 0 {
		return nil, fmt.Errorf("document ID is required")
	}

	doc, err := s.documents.GetByID(ctx, documentID)
	if err != nil {
		return nil, fmt.Errorf("getting document: %w", err)
	}

	if !ocrSupportedContentTypes[doc.ContentType] {
		return nil, fmt.Errorf("document content type %q is not supported for OCR; supported: image/jpeg, image/png, application/pdf", doc.ContentType)
	}

	filePath, contentType, err := s.documents.GetFilePath(ctx, documentID)
	if err != nil {
		return nil, fmt.Errorf("getting document file path: %w", err)
	}

	fileData, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("document file unavailable")
	}

	result, err := s.provider.ProcessImage(ctx, fileData, contentType)
	if err != nil {
		return nil, fmt.Errorf("OCR processing failed: %w", err)
	}

	return result, nil
}

File: rust/crates/zfaktury-core/src/service/ocr_svc.rs
use std::sync::Arc;

use zfaktury_domain::{DomainError, OCRResult};

use super::document_svc::DocumentService;

/// Trait for OCR providers, abstracted for testability.
pub trait OCRProvider: Send + Sync {
    fn process_image(&self, data: &[u8], content_type: &str) -> Result<OCRResult, DomainError>;
}

/// Service for OCR processing of expense documents.
#[allow(dead_code)]
pub struct OCRService {
    provider: Arc<dyn OCRProvider>,
    documents: Arc<DocumentService>,
}

impl OCRService {
    pub fn new(provider: Arc<dyn OCRProvider>, documents: Arc<DocumentService>) -> Self {
        Self {
            provider,
            documents,
        }
    }

    /// Process raw bytes through OCR without requiring a stored document.
    /// Used by views that have file bytes from rfd file picker.
    pub fn process_bytes(&self, data: &[u8], content_type: &str) -> Result<OCRResult, DomainError> {
        let supported = ["image/jpeg", "image/png", "application/pdf"];
        if !supported.contains(&content_type) {
            return Err(DomainError::InvalidInput);
        }
        if data.is_empty() {
            return Err(DomainError::InvalidInput);
        }
        self.provider.process_image(data, content_type)
    }

    /// Processes a stored document by ID through OCR.
    pub fn process_document(&self, document_id: i64) -> Result<OCRResult, DomainError> {
        if document_id == 0 {
            return Err(DomainError::InvalidInput);
        }
        let doc = self.documents.get_by_id(document_id)?;

        let supported = ["image/jpeg", "image/png", "application/pdf"];
        if !supported.contains(&doc.content_type.as_str()) {
            return Err(DomainError::InvalidInput);
        }

        let data = std::fs::read(&doc.storage_path).map_err(|e| {
            log::error!("reading document file {}: {e}", doc.storage_path);
            DomainError::NotFound
        })?;

        self.provider.process_image(&data, &doc.content_type)
    }
}
